src/bluecherry.c

```diff
@@ -439,6 +445,85 @@ static int _bluecherry_dtls_recv(void *ctx, unsigned char *buf, size_t len)
   return recv(sock, buf, len, 0);
 }
 
+static esp_err_t bluecherry_connect(void)
+{
+    int ret;
+    const int handshake_total_timeout_s = 10; /* tune as needed */
+
+    ESP_LOGI(TAG, "Connecting to the BlueCherry platform");
+
+    /* close old socket */
+    if (_bluecherry_opdata.sock >= 0) {
+        shutdown(_bluecherry_opdata.sock, 0);
+        close(_bluecherry_opdata.sock);
+        _bluecherry_opdata.sock = -1;
+    }
+
+    struct addrinfo hints = {0};
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_DGRAM;
+    struct addrinfo *res = NULL;
+    ret = getaddrinfo(BLUECHERRY_HOST, BLUECHERRY_PORT, &hints, &res);
+    if (ret != 0 || res == NULL) {
+        ESP_LOGE(TAG, "DNS lookup failed: %d", ret);
+        if (res) freeaddrinfo(res);
+        return ESP_FAIL;
+    }
+
+    _bluecherry_opdata.sock = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+    if (_bluecherry_opdata.sock < 0) {
+        ESP_LOGE(TAG, "socket() failed: %s", strerror(errno));
+        freeaddrinfo(res);
+        return ESP_FAIL;
+    }
+
+    struct timeval timeout;
+    timeout.tv_sec = 3;
+    timeout.tv_usec = 0;
+    setsockopt(_bluecherry_opdata.sock, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));
+
+    if (connect(_bluecherry_opdata.sock, res->ai_addr, res->ai_addrlen) != 0) {
+        ESP_LOGE(TAG, "connect() failed: %s", strerror(errno));
+        close(_bluecherry_opdata.sock);
+        _bluecherry_opdata.sock = -1;
+        freeaddrinfo(res);
+        return ESP_FAIL;
+    }
+    freeaddrinfo(res);
+
+    /* reset SSL context and clear any previous session */
+    mbedtls_ssl_session_reset(&_bluecherry_opdata.ssl);
+    mbedtls_ssl_set_bio(&_bluecherry_opdata.ssl,
+                        &_bluecherry_opdata.sock,
+                        _bluecherry_dtls_send,
+                        _bluecherry_dtls_recv,
+                        NULL);
+
+    /* handshake with bounded total timeout */
+    time_t t_start = time(NULL);
+    while ((ret = mbedtls_ssl_handshake(&_bluecherry_opdata.ssl)) != 0) {
+        if (ret == MBEDTLS_ERR_SSL_WANT_READ || ret == MBEDTLS_ERR_SSL_WANT_WRITE ||
+            ret == MBEDTLS_ERR_SSL_TIMEOUT) {
+            if (difftime(time(NULL), t_start) >= handshake_total_timeout_s) {
+                ESP_LOGE(TAG, "DTLS handshake timed out");
+                return ESP_ERR_TIMEOUT;
+            }
+
+            esp_task_wdt_reset();
+            vTaskDelay(pdMS_TO_TICKS(10));
+            continue;
+        }
+
+        ESP_LOGE(TAG, "DTLS handshake failed: -0x%04X", -ret);
+        return ESP_FAIL;
+    }
+
+    ESP_LOGI(TAG, "BlueCherry cloud connection established successfully");
```

Comment:
I like to keep the library clean of LOGI/D and only include LOGE. This to keep the footprint as small as possible. It's up to the end user to add extra logs in the application layer.